include/graaflib/algorithm/strongly_connected_components.tpp

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include <graaflib/algorithm/strongly_connected_components.h>
+
+#include <functional>  // For std::function
+#include <stack>
+#include <unordered_map>
+#include <vector>
+
+namespace graaf::algorithm {
+
+template <typename V, typename E>
+[[nodiscard]] std::vector<std::vector<vertex_id_t>>
+tarjans_strongly_connected_components(
+    const graph<V, E, graph_type::DIRECTED>& graph) {
+  // Vector to store strongly connected components
+  std::vector<std::vector<vertex_id_t>> sccs;
+
+  // Stack to hold vertices during traversal
+  std::stack<vertex_id_t> stack;
+
+  // Maps to keep track of indices, low-link values, and stack membership
+  std::unordered_map<vertex_id_t, size_t> indices;
+  std::unordered_map<vertex_id_t, size_t> low_links;
+  std::unordered_map<vertex_id_t, bool> on_stack;
+
+  // Counter for indexing vertices
+  size_t index_counter = 0;
+
+  // Lambda function for the strong connect traversal
+  std::function<void(vertex_id_t)> strong_connect;
+
+  strong_connect = [&](vertex_id_t vertex) {
```

Comment:
[very minor]: Since we are not passing around the lambda in an interface, we can keep it a lambda closure rather than assigning it to a function object:
```suggestion
  // Lambda function for the strong connect traversal
  const auto strong_connect = [&](vertex_id_t vertex) {
```

Reply:
Ah, I see now why you did this, since the function object is used in the body the compiler needs to know the type. Reverted this change in [35036c3](https://github.com/bobluppes/graaf/pull/78/commits/35036c3f6b39e9bca14dd62a9b97c30f350a11e2).

It might also be interesting to see if we can implement this in terms of the existing traversal algorithms in `graph_traversal.h`, but I will leave this for a follow up.